### ISSUE TITLE
(Core) Health / Max Health updates are not being sent to client for some creatures

### ISSUE THREAD

**Issue by kjack9** [azerothcore/azerothcore-wotlk#17655]
### Current Behaviour

Some creatures (exact criteria is yet unknown) are not sending correct updates for health and max health to the client. This results in creatures who appear to have one health/max health value but when inspected with `.npc info` have a different value. Attacking the creature updates the current health but not the max health.

### Expected Blizzlike Behaviour

Server health and client health values should always match.

### Source

Discovered in https://github.com/azerothcore/mod-autobalance/issues/169

### Steps to reproduce the problem

Strap in.

### Setup
1. Fresh AC, no changes
2. Install [https://github.com/kjack9/mod-healthbugtest](this module) to help illustrate the issue as simply as possible. It will start all dungeon creatures with 500 max health (`OnCreatureAddWorld()`) and then set max health to the number of players in the instance * 1000 in `OnAllCreatureUpdate()`.
3. Log in with two different  level 70+ characters and put them into a raid
4. On both Character A and Character B:
    - Turn off GM mode  - `.gm off`
    - Teleport to just outside Gruul's Lair  - `.tele gruulslair`
    - Unbind all instances - `.inst un all`

### Test - One Player, transition from initial MaxHealth (`500`) to updated MaxHealth(`1000`)
1. With Character **A**, step into the instance and do not proceed further in. All the following commands are completed with Character **A**.
    - During map creation, all creatures are spawned with MaxHealth of `500`. As soon as you enter, the first `OnAllCreatureUpdate()` updates their health to `numPlayers * 1000` = `1000`.
2. Target any ogre in front of you
    - Observe that its MaxHealth has been changed to `1000` (correct) and its health is regenerating
3. `/tar Blindeye`  (who is standing next to Kiggler)
    - Observe that Blindeye's MaxHealth has been changed to `1000` (correct)
4. `/tar kiggler`
    - Observe that Kiggler has a displayed Health and MaxHealth of `500` (incorrect)
5. `.npc info`
    - Observe that Kiggler's health according to the server is `1000` current and `1000` max (correct)

The next test assumes that you have completed this one and have not made any further changes.

### Test - Two Players, transition from 1-Player MaxHealth (`1000`) to 2-Player MaxHealth (`2000`)
1. On Character **B**, step into the instance and do not proceed further in
2. On Character **B**, target any ogre in front of you
    - Observe that its MaxHealth has been changed to `2000` (correct) and its health is regenerating
3. On Character **B**, `/tar Blindeye`  (who is standing next to Kiggler)
    - Observe that Blindeye's MaxHealth has been changed to `2000` (correct)
4. On Character **B**, `/tar kiggler`
    - Observe that Kiggler has a displayed Health and MaxHealth of `1000` (incorrect)
5. On Character **B**, `.npc info`
    - Observe that Kiggler's health according to the server is `2000` current and `2000` max (correct)
6. On Character **A**, `/tar kiggler`
    - Observe that Kiggler has a displayed Health and MaxHealth of `500` (incorrect, unchanged)

The next test assumes that you have completed this one and have not made any further changes.

### Test - Two Players, one in close proximity to Kiggler
1. On Character **B**, step out of the instance.
1. On Character **A**, `.go xyz 95.075272 164.496216 -13.491748 565 0.726687`, which will bring you near to where Kiggler is
2. On Character **A**, `/tar kiggler`
    - Observe that Kiggler has a displayed Health and MaxHealth of `500` (incorrect, unchanged)
3. On Character **B**, step into the instance.
4. On Character **A**, `/tar kiggler`
    - Observe that on Character **A**, Kiggler has a MaxHealth of `2000` (correct!)
    - Observe that on Character **B**, Kiggler has a MaxHealth of `1000` (incorrect)
5. Exit the instance with both characters

### Test - One Player, teleport to Kiggler
All actions done with Character **A**
1. `.inst un all`
2. `.go xyz 95.075272 164.496216 -13.491748 565 0.726687`
3. `/tar kiggler`
    - Observe that Kiggler has a MaxHealth of `1000` (correct!) is is regenerating health

### Test - Attacking
All actions done with Character **B**
1. `/inst un all`
2. Enter the instance on foot
3. `/tar kiggler`
    - Observe that Kiggler has a MaxHealth of `1000` (incorrect, should be `2000` for 2 players)
4. `.gm visible off`
5. Walk to Kiggler
    - Observe that Kiggler's displayed health is still wrong
6. Auto-attack Kiggler (you are invisible, he won't hit back)
    - Observe that Kiggler's Health value is now `2000` (correct) but his MaxHealth is still `1000` (decidedly incorrect)

### AC rev. hash/commit

```
AC> .server debug
AC> AzerothCore rev. 4db362492e72 2023-11-07 18:53:27 +0000 (master branch) (Unix, RelWithDebInfo, Static)
Connected players: 2. Characters in world: 2.
Connection peak: 2.
Server uptime: 38 minute(s) 27 second(s)
Update time diff: 1ms, average: 1ms.
Using SSL version: OpenSSL 3.0.2 15 Mar 2022 (library: OpenSSL 3.0.2 15 Mar 2022)
Using Boost version: 1.74.0
Using CMake version: 3.22.1
Using MySQL version: 80034
Found MySQL Executable: /usr/bin/mysql
Compiled on: Linux 5.15.0-87-generic
Worldserver listening connections on port 8085
Realmlist (Realm Id: 2) configured in port 8085
VMAPs status: Enabled. LineOfSight: 1, getHeight: 1, indoorCheck: 1
MMAPs status: Enabled
maps directory located in /azerothcore/env/dist/data/maps. Total size: 291014951 bytes
vmaps directory located in /azerothcore/env/dist/data/vmaps. Total size: 658130721 bytes
mmaps directory located in /azerothcore/env/dist/data/mmaps. Total size: 2192910844 bytes
Default DBC locale: enUS.
All available DBC locales: enUS 
Using World DB: ACDB 335.10-dev
Latest LoginDatabase update: 2023_04_24_00.sql
Latest CharacterDatabase update: trasmorg.sql
Latest WorldDatabase update: 2023_11_07_02.sql
LoginDatabase queue size: 0
CharacterDatabase queue size: 0
WorldDatabase queue size: 0
> List enable modules:
- mod-healthbugtest
```

### Operating system

Ubuntu 22.04

### Custom changes or Modules

Only the test module.

**Comment by kjack9** [comment]
No concrete reason to think so yet, but I would appreciate if @AGandrup would take a look and see if this sounds related to their changes.

**Comment by heyitsbench** [comment]
Would you be able/willing to test the issue on a revision prior to the visibility commit(s)?

**Comment by kjack9** [comment]
> Would you be able/willing to test the issue on a revision prior to the visibility commit(s)?

Absolutely, will get to that tonight.

**Comment by kjack9** [comment]
Looks like this bug existed before the visibility notifier changes. GOOD NEWS @AGandrup! That being said, the latest improvement patch for visibility notifier did change the results slightly

All test done with characters stepping in the instance portal and using `/tar` to target Kiggler

**Last commit before visibility notifier changes started**: 2779833768bbe9967006d6d16888dace05b4bcb6
1-player test: Char A sees health as `500/1000` (should be `1000/1000`), doesn't regenerate. `.npc info` shows correct values.
2-player test: Char A sees `500/1000` (should be 2000/2000). Char B sees `1000/2000`. `.npc info` correct.

**feat(Core/Grids): Implement visibility notifier (#15919)**: 2779833768bbe9967006d6d16888dace05b4bcb6
Same results as 2779833768bbe9967006d6d16888dace05b4bcb6

**fix(Core/Grid): Address bugs and performance issues introduced by visibility notifier implementation (#17480)**: 60e27511c530b0949fee361a89d3eb9d9f9125af
1-player test: Char A sees health as `500/500` (should be `1000/1000`. `.npc info` shows correct values.
2-player test: Char A sees health as `500/500` still (should be `2000/2000`). Char B sees `1000/1000`. `.npc info` correct

**feat(Core/Crypto): add support `OpenSSL 3.0` (#13354)**: a1a1528cb4a9ed6f0220621fe6ecf9e2c36a5534
First commit that supports OpenSSL 3.x so I don't have to install OpenSSL 1.x. Let me know if you want me to go further back.
1-player test: Char A sees health as `500/1000` (should be `1000/1000`), doesn't regenerate. `.npc info` shows correct values.
2-player test: Char A sees `500/1000` (should be 2000/2000). Char B sees `1000/2000`. `.npc info` correct.

TL:DR - this bug has been around a _while_, and would not have been apparent unless the max health of a creature is being changed. 60e27511c530b0949fee361a89d3eb9d9f9125af changed its behavior somewhat, but didn't fix it.

**Comment by kjack9** [comment]
Just realized I forgot to push the changes to the test module I made. Should be updated now.

https://github.com/kjack9/mod-healthbugtest

**Comment by kissingers** [comment]
After update grid, also some new problem about the map ground.
When the Sunwell the first boss is killed, a new boss body is summoned and presented items, but now there is again roughly a 1-2% chance , the body dropping to the lower floor and cant get items.
